test/unit_tests/hoc_python/test_reference_counts.py

```diff
@@ -0,0 +1,281 @@
+    section()
+    assert sys.getrefcount(section) == 2
+
+    # {Py_tp_getattro, (void*) section_getattro_safe},
+    section.L
+    assert sys.getrefcount(section) == 2
+
+    section.Ra
+    assert sys.getrefcount(section) == 2
+
+    section.nseg
+    assert sys.getrefcount(section) == 2
+
+    section.rallbranch
+    assert sys.getrefcount(section) == 2
+
+    # not sure what the purpose of this is; return a dict of
+    # {'L': None, 'Ra': None, 'nseg': None, 'rallbranch': None}
+    d = section.__dict__
+    assert sys.getrefcount(d) == 2
+    assert sys.getrefcount(section) == 2
+
+    # {Py_tp_richcompare, (void*) pysec_richcmp_safe},
+    section == section
+    section != section
+    section < section
+    section <= section
+    section > section
+    section >= section
+    assert sys.getrefcount(section) == 2
+
+    # {Py_tp_iter, (void*) seg_of_section_iter_safe},
+
+    # {Py_tp_methods, (void*) NPySecObj_methods},
+    dir(section)
+    assert sys.getrefcount(section) == 2
+
+    # {Py_tp_doc, (void*) "Section objects"},
+    help(section)
+    assert sys.getrefcount(section) == 2
+
+    # {Py_sq_contains, (void*) NPySecObj_contains_safe},
+    "foo" in section
+    assert sys.getrefcount(section) == 2
+
+    #static PyType_Slot nrnpy_AllSegOfSecIterType_slots[] = {
+    it = section.allseg()
+    assert sys.getrefcount(it) == 2
+    assert sys.getrefcount(section) == 3
+    del it
+
+    # static PyType_Spec nrnpy_SegOfSecIterType_spec = {
+    it = iter(section)
+    assert sys.getrefcount(it) == 2
+    assert sys.getrefcount(section) == 3
+
+    n = next(it)
+    assert sys.getrefcount(it) == 2
+    assert sys.getrefcount(section) == 4
+    del n
+    assert sys.getrefcount(section) == 3
+    del it
+    assert sys.getrefcount(section) == 2
+
+
+def test_segment():
+    section = h.Section(name='section_segment')
+    segment = section(0.5)
+    assert sys.getrefcount(section) == 3
+
+    assert sys.getrefcount(segment) == 2
+
+    # {Py_tp_repr, (void*) pyseg_repr_safe},
+    repr(segment)
+    assert sys.getrefcount(segment) == 2
+
+    # {Py_tp_hash, (void*) pyseg_hash_safe},
+    hash(segment)
+    assert sys.getrefcount(segment) == 2
+
+    # {Py_tp_getattro, (void*) segment_getattro_safe},
+    segment.v
+    assert sys.getrefcount(segment) == 2
+
+    #} else if ((otype = PyDict_GetItemString(pmech_types, n)) != NULL) {
+    #    int type = PyInt_AsLong(otype);
+    #    // printf("segment_getattr type=%d\n", type);
+    #    Node* nd = node_exact(sec, self->x_);
+    #    Prop* p = nrn_mechanism(type, nd);
+    #    if (!p) {
+    #        rv_noexist(sec, n, self->x_, 1);
+    #        return nullptr;
+    #    } else {
+    #        result = (PyObject*) new_pymechobj(self, p);
+    #    }
+
+    #} else if (strncmp(n, "_ref_", 5) == 0) {
+    segment._ref_v
+    assert sys.getrefcount(segment) == 2
+
+    #todo: need to find hoc_built_in_symlist values
+
+    # not sure what the purpose of this is; return a dict of
```

Comment:
This is vital, it's how you discover (some of) the available "things" like mechanisms, or methods (along with some things that aren't supposed to be there).

Reply:
Hmm, that's the last resort, normally I would expect `dir()` to be used, which relies on `Py_tp_methods` and things like `Py_tp_getset`.

Anyways, not important for this review, I will remove the `not sure` as it's not relevant.